File: hllpp_py/src/hllpp/Hasher.py
############################################
# Copyright (c) 2026 Shun/修海 (@shun4midx) #
# Project: HyperLogLogPlusPlus-Autocorrect #
# File Type: Python file                   #
# File: Hasher.py                          #
############################################

import struct

def murmur3_64(key: str, seed: int = 42) -> int:
    key_bytes = key.encode('utf-8')
    length = len(key_bytes)
    m = 0xc6a4a7935bd1e995
    r = 47

    h = seed ^ (length * m)

    num_blocks = length // 8
    for i in range(num_blocks):
        k = struct.unpack_from('<Q', key_bytes, i * 8)[0]  # Little-endian 64-bit
        k *= m
        k ^= k >> r
        k *= m

        h ^= k
        h *= m

    remaining = key_bytes[num_blocks * 8:]
    remaining_val = 0
    for i in range(len(remaining)):
        remaining_val |= remaining[i] << (i * 8)
    if remaining:
        h ^= remaining_val
        h *= m

    h ^= h >> r
    h *= m
    h ^= h >> r

    return h & 0xFFFFFFFFFFFFFFFF  # Return as unsigned 64-bit

def str_to_u64(s: str) -> int:
    return murmur3_64(s)

File: hllpp_cpp/src/src/Hasher.cpp
/********************************************
 * Copyright (c) 2026 Shun/修海 (@shun4midx) *
 * Project: HyperLogLogPlusPlus-Autocorrect *
 * File Type: C++ file                      *
 * File: Hasher.cpp                         *
 ****************************************** */

// ======== INCLUDE ======== //
#include "../include/HLLPP/Hasher.h"

// ======= HASHER ======== //
uint64_t byteswap64(uint64_t x) {
    return ((x & 0x00000000000000FFULL) << 56) |
           ((x & 0x000000000000FF00ULL) << 40) |
           ((x & 0x0000000000FF0000ULL) << 24) |
           ((x & 0x00000000FF000000ULL) << 8)  |
           ((x & 0x000000FF00000000ULL) >> 8)  |
           ((x & 0x0000FF0000000000ULL) >> 24) |
           ((x & 0x00FF000000000000ULL) >> 40) |
           ((x & 0xFF00000000000000ULL) >> 56);
}

uint64_t murmur3_64(const std::string& key, uint64_t seed = 42) {
    const uint64_t m = 0xc6a4a7935bd1e995ULL;
    const int r = 47;

    size_t len = key.size();
    size_t nblocks = len / 8;

    uint64_t h = seed ^ (len * m);

    for (size_t i = 0; i < nblocks; ++i) {
        uint64_t k;
        std::memcpy(&k, key.data() + i*8, sizeof(k));

        // Big Endian support
        #if __has_include(<bit>) && defined(__cplusplus) && __cplusplus >= 202002L
            if constexpr (std::endian::native == std::endian::big) {
                k = byteswap64(k);
            }
        #elif defined(__BYTE_ORDER__) && __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
            k = __builtin_bswap64(k);
        #endif

        k *= m;
        k ^= k >> r;
        k *= m;

        h ^= k;
        h *= m;
    }

    const unsigned char* tail = reinterpret_cast<const unsigned char*>(key.data()) + nblocks * 8;
    uint64_t rem = 0;
    switch (len & 7) {
        case 7: rem |= uint64_t(tail[6]) << 48; [[fallthrough]];
        case 6: rem |= uint64_t(tail[5]) << 40; [[fallthrough]];
        case 5: rem |= uint64_t(tail[4]) << 32; [[fallthrough]];
        case 4: rem |= uint64_t(tail[3]) << 24; [[fallthrough]];
        case 3: rem |= uint64_t(tail[2]) << 16; [[fallthrough]];
        case 2: rem |= uint64_t(tail[1]) <<  8; [[fallthrough]];
        case 1: rem |= uint64_t(tail[0]);
                h ^= rem;
                h *= m;
    }

    h ^= h >> r;
    h *= m;
    h ^= h >> r;
    return h;
}

uint64_t str_to_u64(const std::string& str) {
    return murmur3_64(str);
}